File: src/backend/data/models/catering.js
// data/models/solicitudCatering.js
import { DataTypes } from "sequelize";
import { sequelize } from "../config/db.js";
import Usuario from "./usuario.js";

const SolicitudCatering = sequelize.define(
    "SolicitudCatering",
    {
        idSolicitud: {
            type: DataTypes.INTEGER,
            autoIncrement: true,
            primaryKey: true,
        },
        idUsuario: {
            type: DataTypes.INTEGER,
            allowNull: true,
        },
        nombre_completo: {
            type: DataTypes.STRING(200),
            allowNull: false,
        },
        correo: {
            type: DataTypes.STRING(150),
            allowNull: false,
        },
        telefono: {
            type: DataTypes.STRING(20),
            allowNull: false,
        },
        fecha_evento: {
            type: DataTypes.DATEONLY,
            allowNull: false,
        },
        hora_evento: {
            type: DataTypes.TIME,
            allowNull: true,
        },
        tipo_evento: {
            type: DataTypes.STRING(100),
            allowNull: false,
        },
        cantidad_personas: {
            type: DataTypes.INTEGER,
            allowNull: true,
        },
        descripcion: {
            type: DataTypes.TEXT,
            allowNull: true,
        },
        estado: {
            type: DataTypes.ENUM("pendiente", "confirmada", "rechazada", "cancelada"),
            defaultValue: "pendiente",
            allowNull: false,
        },
        fecha_solicitud: {
            type: DataTypes.DATE,
            defaultValue: sequelize.literal("CURRENT_TIMESTAMP"),
            allowNull: false,
        },
        respondida_por: {
            type: DataTypes.INTEGER,
            allowNull: true,
        },
        respuesta: {
            type: DataTypes.TEXT,
            allowNull: true,
        },
    },
    {
        tableName: "solicitud_catering",
        timestamps: false,
        indexes: [
            { fields: ["idUsuario"] },
            { fields: ["fecha_evento"] },
            { fields: ["estado"] },
            { fields: ["respondida_por"] },
        ],
    }
);

// Relaciones
SolicitudCatering.belongsTo(Usuario, {
    foreignKey: "idUsuario",
    as: "usuario",
});

SolicitudCatering.belongsTo(Usuario, {
    foreignKey: "respondida_por",
    as: "administradorRespuesta",
});

// Relaciones inversas (útiles para consultas)
Usuario.hasMany(SolicitudCatering, {
    foreignKey: "idUsuario",
    as: "solicitudes_catering",
});

Usuario.hasMany(SolicitudCatering, {
    foreignKey: "respondida_por",
    as: "solicitudes_respondidas",
});

export default SolicitudCatering;

File: src/movil/lib/models/catering.dart
// lib/models/catering.dart
class Catering {
  final int idSolicitud;
  final int? idUsuario;
  final String nombreCompleto;
  final String correo;
  final String telefono;
  final String fechaEvento; // 'YYYY-MM-DD'
  final String? horaEvento; // 'HH:mm:ss' o null
  final String tipoEvento;
  final int? cantidadPersonas;
  final String? descripcion;
  final String estado; // pendiente, confirmada, rechazada, cancelada
  final String fechaSolicitud; // fecha + hora
  final int? respondidaPor;
  final String? respuesta;

  // Datos del admin que respondió (opcional, cuando el backend lo incluye)
  final String? nombreAdminRespuesta;

  Catering({
    required this.idSolicitud,
    this.idUsuario,
    required this.nombreCompleto,
    required this.correo,
    required this.telefono,
    required this.fechaEvento,
    this.horaEvento,
    required this.tipoEvento,
    this.cantidadPersonas,
    this.descripcion,
    required this.estado,
    required this.fechaSolicitud,
    this.respondidaPor,
    this.respuesta,
    this.nombreAdminRespuesta,
  });

  factory Catering.fromJson(Map<String, dynamic> json) {
    return Catering(
      idSolicitud: json['idSolicitud'],
      idUsuario: json['idUsuario'],
      nombreCompleto: json['nombreCompleto'] ?? json['nombre_completo'] ?? '',
      correo: json['correo'] ?? '',
      telefono: json['telefono'] ?? '',
      fechaEvento: json['fechaEvento'] ?? json['fecha_evento'] ?? '',
      horaEvento: json['horaEvento'] ?? json['hora_evento'],
      tipoEvento: json['tipoEvento'] ?? json['tipo_evento'] ?? '',
      cantidadPersonas: json['cantidadPersonas'] ?? json['cantidad_personas'],
      descripcion: json['descripcion'],
      estado: json['estado'] ?? 'pendiente',
      fechaSolicitud: json['fechaSolicitud'] ?? json['fecha_solicitud'] ?? '',
      respondidaPor: json['respondidaPor'] ?? json['respondida_por'],
      respuesta: json['respuesta'],
      nombreAdminRespuesta: json['administradorRespuesta']?['username'],
    );
  }
}
